### PR TITLE
Scheduler: fix ConcurrentModificationException

### LINKED ISSUE #3504: [Scheduler] ConcurrentModificationException in ExpressionThreadPoolManager
When testing #3501 the log showed plenty of these:

```
Exception in thread "ESH-astro-7" java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1437)
	at java.util.HashMap$KeyIterator.next(HashMap.java:1461)
	at org.eclipse.smarthome.core.scheduler.ExpressionThreadPoolManager$ExpressionThreadPoolExecutor.afterExecute(ExpressionThreadPoolManager.java:108)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1150)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

```

### PR DESCRIPTION
**Pull request by sjka**
... by expanding the synchronized block to include the outer loop. 

And while being at it anyway I also vaporized the removal logic to a single line.

fixes #3504
Signed-off-by: Simon Kaufmann <simon.kfm@googlemail.com>

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/scheduler/ExpressionThreadPoolManager.java [hunk 1]
```diff
@@ -105,20 +105,9 @@ protected void afterExecute(Runnable runnable, Throwable throwable) {
...
-                                break;
-                            }
-                        }
-                        if (toDelete != null) {
-                            logger.trace("Removing Future '{}' (out of {}) for Runnable '{}'", toDelete,
-                                    futures.get(aRunnable).size(), aRunnable);
-                            futures.get(aRunnable).remove(toDelete);
-                        }
+                synchronized (futures) {
+                    for (Runnable aRunnable : futures.keySet()) {
```
Comment (amitjoy) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/scheduler/ExpressionThreadPoolManager.java]: `entrySet()` in map is faster than `keySet()`

Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/scheduler/ExpressionThreadPoolManager.java]: True for HashMap.  However, 

```java
for (Entry<Runnable, ArrayList<Future<?>>> entry : futures.entrySet()) {
    futures.get(entry.getKey()).removeIf(future -> future == runnable);
}
```

is also less readable. Therefore I thought I leave it as it is.

Comment (amitjoy) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/scheduler/ExpressionThreadPoolManager.java]: I agree - readability is as important as efficiency. 👍

Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/scheduler/ExpressionThreadPoolManager.java]: ...at least for small collections like this 😄

**Pull request merged**